core/src/main/java/io/grpc/internal/Http2ClientStream.java

```diff
@@ -94,28 +94,37 @@ protected Http2ClientStream(WritableBufferAllocator bufferAllocator, int maxMess
   protected void transportHeadersReceived(Metadata headers) {
     Preconditions.checkNotNull(headers, "headers");
     if (transportError != null) {
-      // Already received a transport error so just augment it.
+      // Already received a transport error so just augment it. Something is really, really strange.
       transportError = transportError.augmentDescription(headers.toString());
       return;
     }
-    Status httpStatus = statusFromHttpStatus(headers);
-    if (httpStatus == null) {
-      transportError = Status.INTERNAL.withDescription(
-          "received non-terminal headers with no :status");
-    } else if (!httpStatus.isOk()) {
-      transportError = httpStatus;
-    } else {
-      transportError = checkContentType(headers);
-    }
-    if (transportError != null) {
-      // Note we don't immediately report the transport error, instead we wait for more data on the
-      // stream so we can accumulate more detail into the error before reporting it.
-      transportError = transportError.augmentDescription("\n" + headers);
-      transportErrorMetadata = headers;
-      errorCharset = extractCharset(headers);
-    } else {
+    try {
+      if (headersReceived) {
+        transportError = Status.INTERNAL.withDescription("Received headers twice");
+        return;
+      }
+      Integer httpStatus = headers.get(HTTP2_STATUS);
+      if (httpStatus != null && httpStatus >= 100 && httpStatus < 200) {
+        // Ignore the headers. See RFC 7540 §8.1
+        return;
+      }
+      headersReceived = true;
+
+      transportError = validateInitialMetadata(headers);
+      if (transportError != null) {
+        return;
+      }
+
       stripTransportDetails(headers);
       inboundHeadersReceived(headers);
+    } finally {
```

Comment:
"goto error" makes more sense. To avoid nested ifs is also possible though. (Single layer of curly brace vs two layers using `try-catch`)
```
    Integer httpStatus = headers.get(HTTP2_STATUS);
    if (headersReceived) {
      transportError = Status.INTERNAL.withDescription("Received headers twice");
    } else if (httpStatus == null || httpStatus < 100 || httpStatus >= 200) {
      // Ignore the headers if httpStatus is >=100 and <200. See RFC 7540 §8.1 
      headersReceived = true;
      transportError = validateInitialMetadata(headers);
    }

    if (transportError != null) {
      // Note we don't immediately report the transport error, instead we wait for more data on
      // the stream so we can accumulate more detail into the error before reporting it.
      transportError = transportError.augmentDescription(headers.toString());
      transportErrorMetadata = headers;
      errorCharset = extractCharset(headers);
    } else if (httpStatus == null || httpStatus < 100 || httpStatus >= 200) {
      stripTransportDetails(headers);
      inboundHeadersReceived(headers);
    }
```